### ISSUE TITLE
A way to create scrollspy

### ISSUE THREAD

**Issue by VladShcherbin** [brigade/react-waypoint#112]
Hey, guys. As far as I understand, waypoint creates a span element, which is checked if currently visible. This span is just a line. What if the waypoint is a component (with height say 1000px) and you want to trigger onEnter and onLeave when this component is visible.

Is it somehow possible to do?

**Comment by jmeas** [comment]
So far in my apps I've used the span to generate the scrollspy (my scrollbars behave similarly to the ones in Boostrap, which only requires the single span). But I can imagine situations where one might want a scrollspy that factors in the height of some component.

One way to do this would be to add two waypoints: one at the top of your component, and one at the bottom. I know it's not ideal to do this every time, but you should be able to make a simple HoC component abstraction to abstract it. I've been toying with the idea of making a simple wrapper for my particular use case (even though I don't need height to be accounted for), so that I don't have to use the low-level API that this lib provides.

Also, there's been a lot of discussion about converting this lib itself to a HoC (see: #87 and #41 ). I admit I haven't fully read both of those issues – only skimmed them so far – but I remember it being mentioned that factoring in the height of the contents of the waypoint could be taken into account.

If this interests you, you should read up on those issues and see what's been talked about so far. Again, I haven't read them fully but from what I remember other contributors to this lib would at least consider a refactor to make this a HoC. I'm open to it, too!

Lmk if that helps :v:

**Comment by jmeas** [comment]
Ah, interesting @VladShcherbin . I hadn't considered the problem of the component mounting and extending beyond both boundaries. If there was some way to determine the position of a waypoint on mount, then you could do what you're looking for in a robust way. I haven't thought too hard about doing that, but a quick glance at the API suggests that it might not be possible.

I agree that the most elegant way to handle this would be a lib that functions like your second example.

**Comment by trotzig** [comment]
I too hadn't considered the mounting issue. Thanks for adding that to the discussion. If a waypoint was allowed to have children we could take that scenario into account. I'm happy to accept a PR adding support for this use-case using a regular children approach. If we can come up with a use-case that necessitates a HOC I'd be open for that too. The use-cases I've seen so far however haven't really needed a HOC approach.

**Comment by bySabi** [comment]
Waypoint mounting can be delayed, Ex:

```js
class WaypointLazy extends React.Component {

  state = { render: <span style={{fontSize: 0}} /> };

  componentDidMount() {
    setTimeout(() => this.setState({ render: <Waypoint {...this.props} /> }));
  }

  render() {
    return { this.state.render };
  }
}
```

**Comment by bySabi** [comment]
I'm working on two projects relate to scroll behaviours, both derived from React-waypoint code.

First one is [react-scrollaware](https://github.com/bySabi/react-scrollaware) is generic scroll event handler HoC, is almost complete.

Second is [react-scrolled](https://github.com/bySabi/react-scrolled) a WIP project builded on top of `react-scrollaware`. It will be a library of components and HoC scroll behaviours, will include too a Waypoint implementation.  But like a say is still at early stage and good amount of local code is already not commited.

Any feedback is welcome.

**Comment by jmeas** [comment]
This should be resolved by #152. I'm going to close this issue.

@VladShcherbin , I'm curious to hear if the approach in #152 works for your use case.
